### DISCUSSION TITLE
How to add standalone language to demo?

### DISCUSSION THREAD

**Discussion by invm** [CodinGame/monaco-vscode-api#480] [Q&A] [answered]
I'm trying to support a handful of standalone languages, I'm adding a file with a extension from the standalone package and importing `@codingame/monaco-vscode-standalone-languages`

I tried following [issue](https://github.com/CodinGame/monaco-vscode-api/issues/473) without success and even have a different error in the console:
<img width="725" alt="image" src="https://github.com/user-attachments/assets/39382e68-081a-4a36-95c7-e2af8e70d5cd">

In the ticket above a monarch worker is mentioned but I'm not getting an error about a worker, what am I missing in the demo for it to work? Should I remove services or workers for it to work?

**Comment by CGNonofr** [comment, marked as answer]
In the demo, the textmate and theme service overrides are registered, so standalone monarch languages can't be used anymore

**Reply by CGNonofr** [reply]
And you won't be able to remove only them, because the workbench components require the theme service override

**Reply by invm** [reply]
Thanks, trying a different approach.

**Reply by invm** [reply]
So I've managed to add the syntax highlight with the extension register function, manually importing the tm grammar and language config.

Now I'm trying to work with the lsp like this (based on the issue above):
```js
import { MonacoLanguageClient } from 'monaco-languageclient'
import { WebSocketMessageReader, WebSocketMessageWriter, toSocket, listen } from 'vscode-ws-jsonrpc'
import { CloseAction, ErrorAction, MessageConnection, MessageTransports } from 'vscode-languageclient'

const initLspConnection = (url: string): WebSocket => {
  const webSocket = new WebSocket(url)
  webSocket.onopen = () => {
    const socket = toSocket(webSocket)
    const reader = new WebSocketMessageReader(socket)
    const writer = new WebSocketMessageWriter(socket)
    const languageClient = createLanguageClient({
      reader,
      writer
    })
    languageClient.start()
    reader.onClose(() => languageClient.stop())
  }
  return webSocket
}

const createLanguageClient = (transports: MessageTransports): MonacoLanguageClient => {
  return new MonacoLanguageClient({
    name: 'Terraform Language Client',
    id: 'terraform',
    clientOptions: {
      documentSelector: [
        { scheme: 'file', language: 'terraform' },
        { scheme: 'file', language: 'terraform-vars' }
      ],
      // disable the default error handler
      errorHandler: {
        error: () => ({ action: ErrorAction.Continue }),
        closed: () => ({ action: CloseAction.DoNotRestart })
      },
      workspaceFolder: {
        index: 0,
        name: 'workspace',
        uri: Uri.parse('/workspace')
      }
    },
    // create a language client connection from the JSON RPC connection on demand
    connectionProvider: {
      get: () => {
        return Promise.resolve(transports)
      }
    }
  })
}

const webSocket = initLspConnection('ws://localhost:30003/terraform-ls')

listen({
  webSocket,
  onConnection: (connection: MessageConnection) => {
    connection.listen()
  }
})

``` 

I see the initial connection initialization logged in the ls but nothing else, the `onopen` method on the websocket is never called, am I missing something?

**Reply by CGNonofr** [reply]
The call to `listen` overrides your `onopen` assignment and it has no point since you already do everything by hands

**Reply by invm** [reply]
Thanks again for the amazing support for this and the other package, I removed the listen function invocation and everything works, also, you suggested me to use this module instead of `monaco-lanaguageclient` which I ended up doing as it really gives more. Thanks again.
